How to package react components post

franleplant opened the pull request:
This is still wip but it has about 75% of the progress made and I feel comfortable enough sharing it.

franleplant commented on package.json:
```diff
@@ -24,6 +24,7 @@
     "gatsby-plugin-react-helmet": "^3.1.16",
     "gatsby-plugin-sharp": "^2.3.4",
     "gatsby-plugin-typography": "^2.3.18",
+    "gatsby-remark-autolink-headers": "2.1.21",
```
Added this plugin to make all markdown headers be link-able

franleplant commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,819 @@
+its assets, but a lot of progress has been made and there are actually lots of promising solutions.
+
+We are going to explore how this problem is tackled in the wild,
+we will see case studies and at the end of this post we are going
```
TODO: remove case studies (since I completely remove that part) and instead talk about how we related these techniques to real use cases, in each technique we talk about libs that use them in the wild (which is one of the reasons I remove a dedicated case studies section)

franleplant commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,819 @@
+- make sure you have a single copy of `@emotion/core` in your final Application bundle to enable Theming. This is done by properly managing `peerDependencies`, more on this later.
+- see it working in action in the linked repo.
+
+
```
TODO: do some sort of closing section

franleplant commented:
@IgnacioAmigo, just made some changes and completed some of the missing things. I would say it's now about 90% completed. I want to spend some time giving it a detailed read so that I can spot places where I can improve clarity and saying what I want to say in a more efficient way.

I added a Table of Contents plugin that is optional to each post, but for longer posts such as this I figure is a good thing to have.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+  We explore how to package and bundle the code with Webpack so that it can be consumed appropriately,
+  and we explore how to publish and distribute it to NPM so that it can be easily accessible either
+  via the public NPM repository or a private repository.
+  We also analyse different theming strategies for larger Component Libraries and UI Frameworks.
```
Super nit/: _Analyze_ for US audiences according to this: https://writingexplained.org/analyse-or-analyze-difference

 _You should use analyse with primarily British audiences, and analyze with primarily American audiences. Using the nonstandard variant will be distracting to your readers, and detract from your writing._

franleplant replied:
Will push in the following commits

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+## Introduction
+
+> Check [TLDR](#tldr-how-to-package-and-publish-front-end-libraries-in-2020) if you don't want all
+> details about how and why we reach to the conclusion that that is a valid and desirable solution for the
```
All _the_ details?

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+> problem at hand
+
+There are diverse situation in which we want to create an NPM package that contains
+a [React][1] Component with css and images.
```
nit: _CSS_ instead of _css_

franleplant replied:
fixed all

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+This situation is particularly common inside companies that want to share components,
+and pieces of code in general, between apps usually inside a private NPM repository, but
+it also apply to Open Source software like a single Component package such as a Date Picker
```
minor: It also _applies_.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+This situation is particularly common inside companies that want to share components,
+and pieces of code in general, between apps usually inside a private NPM repository, but
+it also apply to Open Source software like a single Component package such as a Date Picker
+or a Multi Select and also to Component Libraries or UI Frameworks like [React Bootstrap][9],
```
I think the idea here is to say that it happens for components/pieces of code related to those frameworks instead of the frameworks themselves which was my initial impression when reading.

franleplant replied:
your initial impression is correct, this situation applies to how those frameworks are built. Will try to clarify it.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+This problem might seem simple enough but it turns out that it has a lot of nuances and trade-offs.
+
+The main challenge in packaging a [React][1] Component is surprisingly not Javascript,
```
"is surprisingly not Javascript" jajajsd

IgnacioAmigo commented on content/blog/package-react-components/index.md:
````diff
@@ -0,0 +1,776 @@
+```
+
+It turns out that there are a lot of schemes for representing this graph in a Packaged NPM module,
+each with it's own trade-offs. Analysing each alternative is the main purpose of this post.
````
minor: _it's_ should be _its_ because it's the possesive pronoun.

franleplant replied:
so many `it'?s`

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+It turns out that there are a lot of schemes for representing this graph in a Packaged NPM module,
+each with it's own trade-offs. Analysing each alternative is the main purpose of this post.
+
+Unfortunately in early 2020 there still no universal standard way of Packaging a Front End module with all
```
minor: This is missing the verb ("there _is_ still no..." or whatever)

IgnacioAmigo replied:
Also when reading it's weird to see the capital P in Packaging because it came up before and it did not have a capital P IIRC.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+its assets, but a lot of progress has been made and there are actually lots of promising solutions.
+
+We are going to explore how this problem is tackled in the wild by studying
+each technique and providing example real world production ready libraries that
```
Had to read this twice. I think if I had to choose I would go with "example of real world, production ready libraries" to make it easier.

franleplant replied:
examples of real world or example of real world (see new version)

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+We are going to explore how this problem is tackled in the wild by studying
+each technique and providing example real world production ready libraries that
+use them and in the end of this post we are going
```
minor/disregard: Como decís "each technique" en realidad el sujeto es singular, no estoy seguro de que "them" sea correcto acá (al menos que sea algo tipo "each of them").

franleplant replied:
yeash, `each of them` is what I want to express

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+Additionally we are going to explore how each of these packaging schemes enable
+different **Theming** schemes and, at the end, recommend one.
+
+Most of this post is also applicable to any piece of Front End Javascript code that is related
```
This was implied earlier in the intro, but I don't think it's bad that it stays, only that it's pretty much the same (or maybe they can be merged or something).

franleplant replied:
Where did I mention that? can't seem to find it

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+In the context of Front End assets, inlining means to bake those assets directly
+into the Javascript code (bundle) generated by our bundlers or compilers such as Webpack or [Babel][4].
+
+CSS can be inlined without any third party library but to enable all regular CSS features
```
To me it was clear that you can inline CSS but without searching, I did not understand what you meant by "to enable all regular CSS features, a third-party CSS-In-JS library is recommended". Specifically the first part about enabling _regular CSS features_ which sounds a little vague or non-specific. Not saying that's bad, but it made me second-guess myself when I was reading this.

IgnacioAmigo replied:
nit: Also, there are 4 instances of "more on this later" in the text, 2 after a comma and 2 within parentheses. I prefer when it's in parentheses. Personally I think it can  seem a little awkard after a comma.

franleplant replied:
Ok, so I rephrase it differently, let me know what you think

franleplant replied:
Replaced all occurrences of `more on this later` for `(more on this later)`

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+CSS can be inlined without any third party library but to enable all regular CSS features
+a third party **CSS-in-JS** library is recommended such as [Emotion][5] or [Styled-components][6], more on this later.
+Images can be inlined by Webpack's [url-loader][11] in the form of [Data URL][100] with base 64 encoding.
```
nit: I'm used to seeing _Base64_ or _base-64_ instead of _base 64_, but not sure about the correctness.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+CSS can be inlined without any third party library but to enable all regular CSS features
+a third party **CSS-in-JS** library is recommended such as [Emotion][5] or [Styled-components][6], more on this later.
+Images can be inlined by Webpack's [url-loader][11] in the form of [Data URL][100] with base 64 encoding.
+SVG can be inlined by turning an `.svg` file into a [React][1] Component (An SVG file is just a regular HTML tree)
```
Unless I misunderstood, the SVG->React transformation means to manually create a component that returns the SVG tag for a specific SVG, right? I know what a React Component looks like and I know what an SVG looks like, but I'm not 100% sure that what I assumed is correct.

franleplant replied:
yes, that is right

franleplant replied:
something like this

```javascript
export default function MyComp() {
  return <svg><circle r=10 /></svg>
}
```

This will render an svg circle (if I actually wrote this correctly) but the main problem with this is that you need to transform regular svg files into this format, most of the time you cannot simply copy the contents of an svg file into the return statement of a react componet because jsx is html and not xml so things such as xml namespaces and other anotations that regular svg drawing tools put into the svg xml tree need to be manually removed.

Being all that said, this is a really great way of dealing with svg because you can modify the svg at runtime via props and assigning those props to different svg attributes such as paths, sizes, colors etc.

franleplant replied:
maybe I can add this to the post? not sure where, maybe an appendix?

IgnacioAmigo replied:
I think it depends on the target, I guess. Your introcution helped me but maybe something who has at least some experience with React does not need it.

franleplant replied:
Added a link to a post that explores this. In fact there is a lib that automatically turns svg into react

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+The main advantage of inlining is that after you bundle all your code base you end up with
+a single Javascript file that contains all the dependencies it needs to run, even CSS and Images.
+
+This makes _very_ easy to package and publish to [NPM][102] and it makes very easy
```
minor: makes _it_ very easy...

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+to recommend a particular solution to this problem based of my experience.
+
+Additionally we are going to explore how each of these packaging schemes enable
+different **Theming** schemes and, at the end, recommend one.
```
First time I read this, I could imagine what you meant with "theming" but I was unsure if there was any concept I was missing. I think this part could be more accessible if it had some parenthesis that explained it in a very introductory manner, maybe.

franleplant replied:
added some clarification, this is a VERY good observation

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+for **Component Consumers** to _use_ your module.
+
+Using **CSS-in-JS** libraries like [Emotion][5] or [Styled-components][6] makes up for a very
+nice **Run-Time** **Theming** API. Since everything is Javascript your Themes are regular Javascript
```
This clears up some confusion I had with what theming actually meant and that I expressed above.

IgnacioAmigo replied:
nit for reading clarity: I think I would put a comma so that it says "Since everything is Javascript**,** your themes are regular Javascript"

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+they tend to be mostly CSS heavy and with a couple of SVGs at most, both of which are efficiently
+inlined.
+
+Images are a bit more problematic because they are inlined in [base 64][103] making them 30% bigger.
```
Gran dato tbh.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+Images are a bit more problematic because they are inlined in [base 64][103] making them 30% bigger.
+
+In my experience, most Component Libraries do not tend to use an over abundance
```
Tend gets repeated twice in this sentence, it's not bad but it caught my attention when first reading.

IgnacioAmigo replied:
nit: Lo busqué pero overabundance se escribe todo junto.

franleplant replied:
fixed and fixed

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+In my experience, most Component Libraries do not tend to use an over abundance
+of images and if they do they tend to be rather small.
+
+Most high resolution images are dynamic content that are usually dealt by
```
remil nit: I'm pretty sure that because _content_ is an uncountable noun, the person is singular (_most high resolution images are dynamic content that **is** usually_)

franleplant replied:
fixed!

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+The main drawback of inlining is that you turn multiple files that can
+be fetched in parallel by the browser to a single larger file that cannot,
+this will become more noticeable in larger code bases such as Big Component Libraries
```
The fact that "Big Component Libraries" made me search to see if it was a new concept I didn't know of.

franleplant replied:
should I de-capitalize it? or maybe simply italicize it? Want to emphasize it is a strong concept but not necessarily is something that is super well defined such as `react`. Although I have seen people talk about it

franleplant replied:
I replaced it for `big Component Libraries`

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+If you have a big enough library of components you might able to break it
+into independent chunks that can be dynamically imported by Applications,
+allowing more parallelism.
+This is actually not a super far fetched solution since dynamic imports
```
_dynamic imports_ was one of those concepts that I really didn't know and so I had to go and search it a little to understand how they make the situation better.

franleplant replied:
Added a link and text explaining a bit the concept. Good observation

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+This is what you as an Author would publish to [NPM][102].
+
+`bundle.css` will use the image by doing `url(./images/dog.jpg)`, so
+both files need to be served in the following URLs
```
nit: Pretty sure I've read "served on [a URL]" and "served from [a URL]" but never "served in"

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+`public` can be any path with a trivial length, the important part is
+that they are both served from [the same path][106]
+
+Aside from properly document how the files inside your package
```
"properly documenting" instead of "properly document"

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+Aside from properly document how the files inside your package
+relate to each other there is not much else you can do.
+
+You can also use this strategy by allowing access to your module via a CDN,
```
I know what a CDN is, but maybe it is valuable to add parentheses and briefly explaining how that works. I understand that you can't explain everything but I don't think a CDN is necessarily a trivial concept here, where most concepts seem to be about front-end.

franleplant replied:
Added a bit more clarification

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+_Component Authors_ might make the module available via a CDN and that might
+remove the need to fiddle around with CSS and Images but you will need to
+access the library via a Browser Global like the good old days.
```
Browser global = global object/resource that scopes to the whole browser?

franleplant replied:
yeash, added a link to the MDN docs

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+#### As Component Authors
+
+The bundler setup is about the same as with **Not-inline**.
+You wont be able to use [create-react-library][105] but you can probably
```
Won't*

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+The Webpack configuration is actually not that hard and a working example
+will be provided at the end of this post.
+
+The key Webpack configuration is the inclusion of [url-loader][11] to
```
It looks like the last 3 paragraphs start with a variation of "the webpack configuration...". Maybe they can be merged in a way that makes it flow?

franleplant replied:
Simplified a bit more, a reduced to 2 jajajaja

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+For larger Component Libraries you will have a greater experience with writing
+and maintaining CSS. You will avoid completely some CSS exclusive problems such as:
+
+- Global Scope and rule collisions
```
I would put a period at the end of these first 2 bullet points to make it consistent with the last one.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+- Global Scope and rule collisions
+- Specificity (inline-styles have maximum specificity, **CSS-in-JS** libraries simulate this)
+- Style composition, I believe that regular Javascript object composition is better than CSS cascading global composition model, plus you can create a custom system that fits your team and your business logic, remember, it's regular Javascript. You can easily publish your Themes as private NPM packages internally to shared them among Applications.
```
nit: Maybe this last bullet point might have too many commas? The first comma can be a colon and maybe the second-to-last can be parenthesis or something like that I think.

franleplant replied:
done

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+need features that inline-styles do not provide and that regular CSS do such
+as pseudo selectors like `:hover` and media queries.
+To solve this problem and provide better Developer Experience overall is that
+**CSS-in-JS** exist.
```
_Exists_ instead of _exist_, I think, if CSS-in-JS is a singular noun.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+That is all!
+
+If you want to import the module dynamically (to parallelize and deffer its loading
```
_Defer_ instead of _deffer_

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+If you have Webpack configured correctly to create chunks that can be
+dynamically loaded, this is called `Code Splitting`, then `module`
+will be an independent file lazy loaded, only and when its needed.
```
_It's_ instead of _its_ (also I think "only and when" sounds a little strange IMHO)

franleplant replied:
NSHO (not so humble opinion).

Fixed!

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+In my experience Webpack configurations are something that most developers
+do not know or do not want to touch so
+**having these sort of abstractions inside
```
_These sorts_ or _this sort_, because the first word refers directly to the second one so it depends on the noun which in the first case is plural and in the second one, singular.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+which translates to less operation costs**.
+These Webpack abstractions have lots of smart people working on them and
+pouring their past experiences in creating something that will cover
+lots of use cases you probably wont ever think of, so you are getting even
```
Won't*

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+need to maintain and bug fix your hand made configurations if you do them yourself.
+It is effectively one more source of bugs you will have to deal with,
+so rely as much as possible on third party open source solutions and for that
+you _must_ avoid **Passive Compilation**.
```
Without knowing much about Webpack, passive compilation sounded pretty nice to me. Letting me decide what I do is valuable, but this section makes a clear point on why it might be a bad idea (though reducing dependencies remains more important). Good section. Is there any case when using passive compilation makes sense?

franleplant replied:
Maybe I need to enforce the point that is actually bad. 

Imagine having different c libraries that all have different sometimes overlapping and sometimes diverging compilation configurations and to make them all run together (instead of using a binary already compiled by the author) you need to accommodate your project to the needs of all those libraries (the example is a bit stretched since i believe there aren't this many discrepancies between compilation configurations in c, c++ and even none in rust), but in the _modern_ js world that is a nightmare.

Also, didn't get the point about reducing dependencies, the amount of dependencies remains about the same, they shift from the author of the component library to the consumer of the component library and contrary to empowering the consumer is just dragging it down. And as I discuss elsewhere, this will infect the rest of your ecosystem and will create a very blurry barrier between apps and components and will tight couple all of these.

franleplant replied:
So, this actually does not give more control to the consumer, it shifts responsibilities

IgnacioAmigo replied:
True, true. Very clear. Regarding reinforcing the points in the main post, I think your second paragraph here might be a good addition but not really necessary.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+##### Disadvantages
+
+You must either specify or follow a specification of what Webpack loaders,
+loader configuration, CSS configuration, Babel configuration, etc your library
```
ultra nit: I think you wrote "etc." earlier instead of "etc"

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+You must either specify or follow a specification of what Webpack loaders,
+loader configuration, CSS configuration, Babel configuration, etc your library
+requires in order for Apps to compile it successfully. You could also construct
+some sort of webpack composition hook but that can get really complicated really
```
"Webpack" instead of "webpack" for consistency.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+simply let _Component Consumers_ use any of your library files
+directly i.e. `import something from 'A/src/some/deep/nested/file.js'`.
+
+This will make even harder to migrated your library to an **Actively Compiled**
```
_Migrate_ instead of _migrated_?

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+This will make even harder to migrated your library to an **Actively Compiled**
+strategy down the road because _Component Consumer_ Apps will be tightly coupled
+to your library's internal file structure and providing that same per-file API
+in Webpack is not trivial and has caveats (sharing code and such).
```
tru, tru

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+**Passively compiled** dependencies to make it work.
+
+If this seems crazy to you I can already tell you that I have
+experienced this in my day to day Job and is super frustrating.
```
Great section as well. I think the previous one (Active vs Passive) is focused more about why not to use passive than comparing them objectively, but some of those disadvantages leak here.

franleplant replied:
That was the intentional spirit, I want to be opinionated here and provide a barrier for people trying to do this (because it might be a super easy and viable solution for starters) and make them reconsider jajajaja.

I know this by heart (unfortunately)

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+#### As Component Authors
+
+You actually need to write a full web pack configuration
```
Webpack? Or did you mean web pack? (Not sure if there's a distinction).

franleplant replied:
Webpack yes

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+for Component Consumers to change the default look of the components to fit
+their brand or their business.
+
+Lets inspect different strategies of providing Theming and Customization hooks.
```
Let's*

franleplant commented:
@IgnacioAmigo I think I addressed all your comments. Good review TBH. I also tweaked the first section's title and the meta description to be more SEO friendly.

I also added a Post Publish checklist to start designing our process a bit better

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+#### As Component Consumer
+
+You use the Component Author provided class names to alter the component looks.
```
_the way the component looks_? or maybe _the component's looks_?

IgnacioAmigo replied:
This also happens above ("The basic premise is that Consumers use well defined CSS class names to alter the Component looks.")

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+The basic premise is that Consumers use well defined CSS class names
+to alter the Component looks.
+
+This approach works only at **Run-Time**.
```
What does it really mean that it works only at runtime and not at build time?

IgnacioAmigo replied:
never mind, I can see the distinction of this versus using CSS/LESS/SASS variables which have to be at build-time (after reading the following sections).

franleplant replied:
I though about having an appendinx or some place else where I talk about build time vs run time, do you think I should reconsider adding it?

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+- [Bootstrap][9]
+- [Semantic-UI][14]
+
+This approach works only at **Build-Time** and **may require a build system**.
```
The "May require" comes from the fact that CSS variables do not require a build system, whereas LESS and SASS variables do require some build system to work?

franleplant replied:
yush

franleplant replied:
Added clarification and links

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,776 @@
+
+#### As Component Consumers
+
+- you will need to either
```
nit: Capital letters to start the bullet points for consistency?

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,814 @@
+
+#### As Component Consumers
+
+- You will need to either
```
nit: Colon after the first bullet point maybe?

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,814 @@
+
+_Out of band_ and _in band_ are terms I got from signal processing theory and in this case I'm using it to
+describe whether you build the StyleSheet in a separate build process and codebase (out of band)
+or you integrate the necessary build configurations into your own build process and codebase (in band)
```
I think the paragraph is missing a period at the end.

IgnacioAmigo replied:
Great section, to the point and accessible.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,814 @@
+This approach uses Core Theme Variables defined at Run-Time and
+passed through Javascript to components to alter their looks.
+
+It works well with the **Inline** Paradigm
```
Missing a period at the end of the sentence.

IgnacioAmigo commented on content/blog/package-react-components/index.md:
```diff
@@ -0,0 +1,814 @@
+#### As Component Authors
+
+- Carefully abstract the core theme variables and derive the rest of the styles from it.
+- Use CSS-in-JS libraries such as Emotion or styled-components
```
Period at the end of the bullet points for constitency.

franleplant commented:
I tried adding Run-Time vs Build-Time but it didn't sell on me. Perhaps if we have questions we might add it later or cover it in a quick post down the road.

I have one more thing TO DO regarding the code examples and after that I will be finished with my part.

franleplant commented:
@IgnacioAmigo I changed my mind about the Passive vs Active compilation section and moved it to be an Appendix, I think it is sort of niche thing but I do want to mention it, I think it will help with the flow of the post since it was like a super long de-tour.
